README.md
<p align="center">
  <img src="https://i.imgur.com/sJzfZsL.jpg" width="154">
  <h1 align="center">InstaPy</h1>
  <p align="center">Tooling that <b>automates</b> your social media interactions to “farm” Likes, Comments, and Followers on Instagram
Implemented in Python using the Selenium module.<p>
  <p align="center">
    <a href="https://github.com/timgrossmann/InstaPy/blob/master/LICENSE">
      <img src="https://img.shields.io/badge/license-GPLv3-blue.svg" />
    </a>
    <a href="https://github.com/SeleniumHQ/selenium">
      <img src="https://img.shields.io/badge/built%20with-Selenium-yellow.svg" />
    </a>
    <a href="https://www.python.org/">
    	<img src="https://img.shields.io/badge/built%20with-Python3-red.svg" />
    </a>
    <a href="https://www.github.com/timgrossmann/InstaPy#backer">
	<img src="https://opencollective.com/instapy/backers/badge.svg">
    </a>
    <a href="https://www.github.com/timgrossmann/InstaPy#sponsors">
	<img src="https://opencollective.com/instapy/sponsors/badge.svg">
    </a>  
    <a href="https://discord.gg/FDETsht">
	<img src="https://img.shields.io/discord/510385886869979136.svg">
    </a>
  </p>
</p>

[Twitter of InstaPy](https://twitter.com/InstaPy) | [Discord Channel](https://discord.gg/FDETsht) | [How it works (FreeCodingCamp)](https://www.freecodecamp.org/news/my-open-source-instagram-bot-got-me-2-500-real-followers-for-5-in-server-costs-e40491358340/) |   
[Talk about automating your Instagram](https://youtu.be/4TmKFZy-ioQ) | [Talk about doing Open-Source work](https://www.youtube.com/watch?v=A_UtST302Og&t=0s&list=PLa4P1NPX9hthXV-wko0xyxFpbhYZFkW7o) | [Listen to the "Talk Python to me"-Episode](https://talkpython.fm/episodes/show/142/automating-the-web-with-selenium-and-instapy)


**Newsletter: [Sign Up for the Newsletter here!](http://eepurl.com/cZbV_v)**   
**Guide to Bot Creation: [Learn to Build your own Bots](https://www.udemy.com/course/the-complete-guide-to-bot-creation/?referralCode=7418EBB47E11E34D86C9)**    

<br />

# Find the full documentation in [Docs](/docs)
**Table of contents**
- [How to install and run InstaPy](/docs/home.md#installation)
  * [Installing InstaPy](/docs/home.md#installation)
  * [Running Instapy](/docs/home.md#installation)
  * [Updating InstaPy](/docs/home.md#updating-instapy)
  * [Guides and tutorials](/docs/home.md#guides)
    * [Video tutorials](/docs/home.md#video-tutorials)
    * [Written guides](/docs/home.md#written-guides)
- [Externals and additionals tools](/docs/home.md#external-tools)
- [Running InstaPy on Docker](/docs/home.md#docker)
- [Documentation of all Instapy's features](/docs/settings.md)
- [Support](/docs/home.md#support)
- [Credits](/docs/home.md#credits)

<br />

## Credits
### Community
An active and supportive community is what every open-source project needs to sustain. Together we reached every continent and most of the countries in the world!   
Thank you all for being part of the InstaPy community ✌️

<p align="center">
	<img src="https://i.imgur.com/XkxHcM7r.png" alt="InstaPy reach" width="500px"/>
</p>

### Contributors

This project exists thanks to all the people who contribute. [[Contribute](https://github.com/timgrossmann/InstaPy/wiki/How-to-Contribute)].

<a href="https://github.com/timgrossmann/InstaPy/graphs/contributors"><img src="https://opencollective.com/instapy/contributors.svg?width=890&button=false" /></a>

### Backers

Thank you to all our backers! 🙏 [[Become a backer](https://opencollective.com/instapy#backer)]

<a href="https://opencollective.com/instapy#backers" target="_blank"><img src="https://opencollective.com/instapy/backers.svg?width=890"></a>

---

> **Disclaimer**<a name="disclaimer" />: Please note that this is a research project. I am by no means responsible for any usage of this tool. Use it on your behalf. I'm also not responsible if your accounts get banned due to the extensive use of this tool.


CHANGELOG.md
# Changelog

The **goal** of this file is explaining to the users of our project the notable changes _relevant to them_ that occurred _between_ commits.

_The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/), and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html)_

## [0.6.17] - 2022-01-19 - [Unreleased]

### Fixed

- Fixed project homepage to [InstaPy](https://github.com/InstaPy/InstaPy)

## [0.6.16] - 2021-12-11

### Added

- Added filter in `extract_text_from_element()` for non-username elements.

### Fixed

- Fixed `is_private_profile()` when 'NoneType' object is not subscriptable
- Fixed `find_element*()` to `find_element()`

## [0.6.16] - 2021-12-11

### Added

- Added filter in `extract_text_from_element()` for non-username elements.

### Fixed

- Fixed `is_private_profile()` when 'NoneType' object is not subscriptable
- Fixed `find_element*()` to `find_element()`

## [0.6.15] - 2021-12-02

### Fixed

- Fixed `relationship_tools` to check if not `username == self_username`

## [0.6.14] - 2021-10-18

### Added

- Added delays after load cookie in login.util.py
- Added `apidisplaypurposes about` in `smart_hashtags` and new api token; Python3.5
- Added `interact_user_likers` interacting with the likers from a given user's posts
- Added second `accept_igcookie_dialogue` to handled the second "cookie accept screen" that is not automatically accepted
- Added delays after `update_activity()` in `follow_util.py` and `util.py`
- Added extra trace in `get_followers()`
- Added `verified_only` argument to `get_followers` to return only verified users
- `mandatory_bio_keywords` control before evaluating for mandatory
- Fix user following and verification of follow
- Skipping of public account with (optional) percentage

### Fixed

- Move call to `get_following_status` above `web_address_navigator` inside `get_links_for_username` function
- Fixed the correct URL `displaypurposes.com`, the old `displaypurposes.com` API is deprecated
- Fixed `additional-information.md`, removed extra info for `cron` configuration
- Fixed `actions.md` and `settings.md` for identation
- Fixed getting additional data from page source
- Fixed xpath for when images cannot be detected
- Fixed empty shared data
- Cannot detect post media type
- Updated xpath Image "Like" and "Unlike"

### Changed

- Change the ip data supplier to <https://freegeoip.app/> from <https://seeip.org/>

## [0.6.13] - 2020-12-30

### Added

- Handle situation if stale elements detected when cookie cannot be loaded
- Convert Documentation to Docusaurus
- Check if user's account is private
- Added informational message when videos found instead of images in `likers_from_photo`
- Two Factor Authentication check when acct is protected
- Added Pull Request Template
- Added function `accept_igcookie_dialogue` that clicks accept at IG cookie dialogue
- Added `comment_liked_photo` to `set_do_comment` to be able to comment twice a photo that is already liked; deactivated by default

### Fixed

- Fixed typo in message when a video is tried to like.
- Fixed the problem where `followers_list` could be used without being initialized.
- Fixed incorrect xpath used by `interact_by_comments`
- Fixed typo in `dissmiss_save_information` to `dismiss_save_information`
- Fixed `get_followers` return for empty list
- Fixed xpath for `likers_from_photo`
- Changed windows size to `iPhone XS Max: 414, 896`
- Fixed `process_comments` where some comments used the InstaPy user
- Updated `user_agent` profile to be the same for the bot session
- Fixed required `sameSite:Strict` parameter in browser cookie session
- Fixed `login_session` for issues when cookie cannot be loaded
- Fixed `likes_dialog_close_xpath` to generic format
- Fixed `validate_username` to navigate user profile to read their Bio
- Fixed `Liked button now found, might be a video` that could be floodig general log
- Fixed `logger.warn` to `logger.warning` as default warning in Python
- Fixed xpath for `watch_story`

## [0.6.12] - 2020-10-26

### Added

- Defined local variables ( ie: 'unfollow_state') to fix message for referenced before assignment
- *.code-workspace in .gitignore for vscode project (macOS)
- `like_by_locations()` with randomize flag
- Added firefox_options.log.level = geckodriver_log_level for debugging
- `dissmiss_save_information` in `login_user()`
- new path used in login_util.py
- information for geckodriver_log_level, development only

### Changed

- check if liking activity was blocked after every like
- Call scroll_down() after follower accepted, with this new def the window is scrolled down after accepting a follower.
- Sorted imports, to make code readable
- Modified followed.txt log to append information
- `like_image` now needs to handle images and videos
- Imports were ordered
- `like_by_feed` limits from 100 to 50

### Fixed

- fixed some comments were double quotes were not necessary
- issue where `getTitle` was timing out
- `get_users_through_dialog_with_graphql` with a new try/except

## [0.6.11] - 2020-09-25

### Added

- Use random tag list for `session.like_by_tags`

### Changed

- General log rotation, gecko driver log in user directory, comments in 80 chars

### Fixed

- Unfollowing of users that haven't posted anything
- `get_links` xpath for yet another change
- Path for Obtaining user id
- `like_util.py` list index out of range error
- `like_by_feed()` method
- `follow_user_following
- processing unexpected alert error on resize

## [0.6.10] - 2020-07-30

### Added

- Generallize mandatory words and add mandatory_bio_keywords

### Changed

- Update xpath for like/unlike and comment
- Fix `like_by_feed()` xpath
- `get_like_on_feed()` improve function readability

### Fixed

- "UnboundLocalError: local variable 'commenting_approved' referenced before assignment" error when bot tries to comment
- Typo updating configuration object. Changed nofity into notify

start.py
import os
import yaml
import random
from instapy import InstaPy
from instapy.util import smart_run

# Importing datafile
current_path = os.path.abspath(os.path.dirname(__file__))
data = yaml.safe_load(open('%s/data.yaml' %(current_path) ))

skip_friends_interaction = data['skip']['friends_interaction']
skip_hashtags_interaction = data['skip']['hashtags_interaction']
skip_locations_interaction = data['skip']['locations_interaction']
skip_follow_unfollow = data['skip']['follow_unfollow']

# Setting data
creds = data['creds']
hashtags = data['tags']
exceptions = data['exceptions']
friends = data['friends']
brands = data['brands']
interests = data['interests']
artists = data['artists']
netherlands = data['netherlands']

# Comments
constructors = [
  u'🔥',
  u'✨',
  u'⚡️',
  u'🙌🏼',
  u'😻',
  u'💪🏼',
  u'😎',
  u'♥️',
  u'💫'
]

comments = []

for x in range(20):
  comment = []
  comment = random.sample(constructors, random.randint(1,5))
  comments.append(u''.join(comment))

# print('Friends Interaction')
# ###############################
# # Interact with close friends #
# # - likes
# if skip_friends_interaction:
#   print('⚡️ Skipping liking friends posts')
# else:
#   print('❤️ Showing friends some love')
#   friend_list = InstaPy(username=creds['user'], password=creds['pass'],
#                         selenium_local_session=False, disable_image_load=True)
#   friend_list.set_selenium_remote_session(selenium_url='http://selenium:4444/wd/hub')
#   friend_list.login()
#   friend_list.set_skip_users(skip_private=False)
#   friend_list.set_relationship_bounds(enabled=True, potency_ratio=0.4)
#   friend_list.set_do_like(True, percentage=80)
#   friend_list.interact_by_users(friends, amount=2, randomize=False)
#   friend_list.end()

# print('🌋 ⛏  📊')
# #########################################
# # Interact with high profiles by hashtag #
# # - likes
# # - comments
# bot = InstaPy(username=creds['user'], password=creds['pass'],
#               selenium_local_session=False, disable_image_load=True)
# bot.set_selenium_remote_session(selenium_url='http://selenium:4444/wd/hub')

# # Don't bother friends
# bot.set_dont_include(friends)

# # Don't get banned 
# bot.set_quota_supervisor(enabled=True, sleep_after=['likes', 'comments_d', 'follows', 'unfollows', 'server_calls_h'],
#                         sleepyhead=True, stochastic_flow=True, notify_me=True, peak_likes=(57, 585),
#                         peak_comments=(21, 182), peak_follows=(48, None), peak_unfollows=(35, 402),
#                         peak_server_calls=(None, 4700))

# # Login
# bot.login()

# # Exceptions
# bot.set_dont_like(exceptions)

# # Comment
# bot.set_comments(comments)
# bot.set_do_comment(enabled=True, percentage=50)

# if skip_hashtags_interaction:
#   print('⚡️ Skipping interaction based on hashtags')
# else:
#   print('🏗 Building hashtags network')
#   # Smart hashtags
#   bot.set_smart_hashtags(hashtags, limit=4, sort='top', log_tags=True)

#   # Setting bounds
#   bot.set_relationship_bounds(enabled=True, potency_ratio=1.5, delimit_by_numbers=True,
#                               max_followers=99999999, max_following=5000,
#                               min_followers=700, min_following=0, min_posts=10)

#   # Like
#   bot.like_by_tags(amount=20, use_smart_hashtags=True)


# ########################
# # Interact by location #
# # - likes
# if skip_friends_interaction:
#   print('⚡️ Skipping interaction based on locations')
# else:
#   print('🏗  Building locations network')
#   # Netherlands
#   bot.like_by_locations(netherlands, amount=2)


# ######################################
# # Follow/Unfollow flow plus interact  #
# # - likes
# # - comments
# if skip_follow_unfollow:
#   print('⚡️ Skipping follow/unfollow')
# else:
#   print('🏔⛏ Mining followers')
#   bot.set_relationship_bounds(enabled=True, potency_ratio=1, delimit_by_numbers=True,
#                               min_followers=250, min_following=50, min_posts=30)

#   ## Don't be an asshole config ##

#   # Setting blacklist
#   bot.set_blacklist(enabled=True, campaign='blacklist')

#   # Dont bother nice people
#   bot.set_dont_include(friends+brands+interests+artists)

#   # Exclude private or accounts without profile picture
#   bot.set_skip_users(skip_private=True, private_percentage=100,
#                     skip_no_profile_pic=True, skip_business=False,)

#   # Set user interaction
#   bot.set_do_like(enabled=True, percentage=100)
#   bot.set_do_comment(enabled=True, percentage=100)
#   bot.set_comments(comments)

#   # Set how much interaction
#   bot.set_user_interact(amount=1, randomize=False, percentage=50)
#   bot.set_simulation(enabled=True)

#   ## Group Nº1 ##
#   # FOLLOW
#   # brands
#   bot.follow_user_followers(brands, amount=100, randomize=False, interact=True, sleep_delay=600)

#   # UNFOLLOW
#   # People that didn't followed back in the last 48hs
#   bot.set_blacklist(enabled=False, campaign='blacklist')
#   bot.unfollow_users(amount=90, InstapyFollowed=(True, 'nonfollowers'), style='FIFO', unfollow_after=12*60*60, sleep_delay=601)

#   ## Group Nº2 ##
#   # FOLLOW
#   # interests+artists
#   bot.set_blacklist(enabled=True, campaign='blacklist')
#   bot.follow_user_followers(interests+artists, amount=100, randomize=False, interact=True, sleep_delay=600)

#   # UNFOLLOW
#   # People that didn't followed back in the last 48hs
#   bot.set_blacklist(enabled=False, campaign='blacklist')
#   bot.unfollow_users(amount=90, InstapyFollowed=(True, 'nonfollowers'), style='FIFO', unfollow_after=12*60*60, sleep_delay=601)

#   # CLEANUP
#   bot.unfollow_users(amount=500, InstapyFollowed=(True, 'all'), style='FIFO', unfollow_after=24*60*60, sleep_delay=601) 

# # Closing sessions
# bot.end()


